File: common/src/main/java/org/sqlbuilder/common/Module.java
package org.sqlbuilder.common;

import java.io.File;
import java.io.FileInputStream;
import java.util.Properties;

public abstract class Module
{
	public final String id;

	protected final Properties props;

	protected Module(final String id, final String conf)
	{
		this.id = id;
		props = getProperties(conf);
	}

	abstract protected void run();

	@Nullable
	public static Properties getProperties(final String conf)
	{
		File confFile = new File(conf);
		try (FileInputStream fis = new FileInputStream(confFile))
		{
			Properties props = new Properties();
			props.load(fis);
			return props;
		}
		catch (Exception ignored)
		{
		}
		return null;
	}
}

File: common/src/main/java/org/sqlbuilder/common/Module.kt
package org.sqlbuilder.common

import java.io.File
import java.io.FileInputStream
import java.util.*

abstract class Module protected constructor(val id: String?, conf: String, @JvmField protected val mode: Mode?) {

    enum class Mode {
        PLAIN, RESOLVE, UPDATE, EXPORT;

        companion object {

                    fun read(arg: String): Mode {
                return when (arg) {
                    "-resolve" -> RESOLVE
                    "-update"  -> UPDATE
                    "-export"  -> EXPORT
                    else       -> PLAIN
                }
            }
        }
    }

    @JvmField
    protected val props: Properties = getProperties(conf)

    protected abstract fun run()

    companion object {

        @Nullable
        fun getProperties(conf: String): Properties {
            val confFile = File(conf)
            FileInputStream(confFile).use {
                val props = Properties()
                props.load(it)
                return props
            }
        }
    }
}
